engine/crates/engine-v2/src/request/walkers/operation_limits.rs

```diff
@@ -0,0 +1,87 @@
+use std::collections::HashSet;
+
+use schema::StringId;
+
+use super::{BoundSelectionSetWalker, BoundSelectionWalker};
+
+impl<'a> BoundSelectionSetWalker<'a, ()> {
+    pub(crate) fn max_depth(&self) -> u16 {
+        (*self)
+            .into_iter()
+            .map(|selection| match selection {
+                BoundSelectionWalker::Field(field) => {
+                    field
+                        .selection_set()
+                        .map(|selection_set| selection_set.max_depth())
+                        .unwrap_or_default()
+                        + 1
+                }
+                BoundSelectionWalker::InlineFragment(inline) => inline.selection_set().max_depth(),
+                BoundSelectionWalker::FragmentSpread(spread) => spread.selection_set().max_depth(),
+            })
+            .max()
+            .expect("must be defined")
+    }
```

Comment:
Would these functions be nicer if they'd take `self` by-value instead of a reference? That would make using them much easier too, as long as `Self: Copy`.

Reply:
But `Self` itself isn't `Copy,` `*self` invokes the `Deref` implementation, and the result of that implement `IntoIterator`. `BoundSelectionSetWalker` isn't supposed to be `Copy`. But @Finistere would know more. 🙂

Reply:
No he's right, it is `Copy` and we could use `self`. Sounds so obvious I don't know why I missed it

Reply:
@Finistere @pimeys ah, okay, sorry, I stand corrected